src/humancellatlas/data/metadata/helpers/staging_area.py

```diff
@@ -0,0 +1,294 @@
+                    file_manifest = self.descriptors[file_id].to_manifest_entry
+                    file_name = file_manifest['name']
+                    if file_name.startswith(f'{subgraph_id}/'):
+                        file_manifest['name'] = file_name.partition('/')[-1]
+                else:
+                    file_manifest = {
+                        'content-type': 'application/json;',
+                        'crc32c': '0' * 8,
+                        'indexed': True,
+                        'name': json_file_name,
+                        's3_etag': None,
+                        'sha1': None,
+                        'sha256': '0' * 64,
+                        'size': len(json.dumps(json_content)),
+                        'uuid': metadata_file.uuid,
+                        'version': metadata_file.version
+                    }
+                manifest.append(file_manifest)
+
+        logger.debug('Bundle contents: %s', list(metadata.keys()))
+        return links_file.version, manifest, metadata
+
+
+@dataclass(frozen=True)
+class GitHubStagingAreaFactory:
+    repo: Repository
+    branch: str
+    path: Tuple[str, ...]
+
+    @classmethod
+    def from_owner_and_name(cls,
+                            owner: str,
+                            name: str,
+                            branch: str,
+                            path: Optional[str] = None
+                            ) -> 'GitHubStagingAreaFactory':
+        """
+        :param owner: The owner of the GitHub repository
+        :param name: The name of the GitHub repository
+        :param branch: A branch name or commit SHA
+        :param path: The path inside the repository to the base of the staging area
+        """
+        token = os.environ['GH_API_TOKEN']  # A GitHub personal access token
+        logger.debug('Requesting GitHub repo %s', (owner, name, branch, path))
+        github_api = Github(token)
+        rate_limit = github_api.get_rate_limit()
+        logger.debug('GitHub API rate limit limit=%i, remaining=%i, reset=%s',
+                     rate_limit.core.limit, rate_limit.core.remaining, rate_limit.core.reset)
+        repo = github_api.get_repo(f'{owner}/{name}')
+        if len(branch) == 40 and set(branch).issubset(set('0123456789abcdef')):
+            branch_sha = branch
+        else:
+            gh_branch = repo.get_branch(branch)
+            branch_sha = gh_branch.commit.sha
+            logger.debug('Branch SHA %s', (branch, branch_sha))
+        path = () if path == '' else tuple(path.split('/'))
+        return cls(repo=repo, branch=branch_sha, path=path)
+
+    @classmethod
+    def from_url(cls, url: str) -> 'GitHubStagingAreaFactory':
+        """
+        :param url: A URL of a staging area in a GitHub repository, specified
+        in 'https://github.com/<OWNER>/<NAME>/tree/<BRANCH>[/<PATH>]' format.
+        """
+        parsed_url = furl(url)
+        assert parsed_url.scheme == 'https', url
```

Comment:
If you use assert here you should use assert instead of `raise ValueError` above.

Or better, copy `require` and `reject` from Azul we'll deal with the duplication later.